### PR TITLE
6LoWPAN cleanup

### PR DESCRIPTION
**Pull request by gpotter2**
Seriously cleanups up the 6lowpan layer:
- it has been the ugliest layer we've had for quite some time
- some bindings were all wrong: fix https://github.com/secdev/scapy/issues/2748
- many weird functions can be reimplemented using things like `MultipleTypeField`
- many fields have funky names ("Destiny address" (?!!! what does it have to do with destiny..), underscores everywhere...)
- add full LOWPAN_HC1 support
- support for LoWPAN_NHC IPv6 extensions, proper support for the UDP one (it was a mess)
- finish some decompression that was left unimplemented
- add links to the doc for each layer because this layer isn't easy to catch on

Test based on the following file:
[6lowpan_nhc.pcap.zip](https://github.com/secdev/scapy/files/5047627/6lowpan_nhc.pcap.zip)

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou @p-l- This is also ready for review

File: scapy/layers/sixlowpan.py [hunk 1]
```diff
@@ -102,96 +125,255 @@ def getfield(self, pkt, s):
...
+        BitEnumField("hc2", 0, 1, ["No more header compression bits",
+                                   "HC2 Present"]),
+        # https://tools.ietf.org/html/rfc4944#section-10.2
+        ConditionalField(
+            MultipleTypeField(
+                [
+                    (PacketField("hc2Field", LoWPAN_HC2_UDP(),
+                                 LoWPAN_HC2_UDP),
+                        lambda pkt: pkt.nh == 1),
+                    # (PacketField("hc2Field", LoWPAN_HC2_ICMP(),
```
Comment (guedou) [inline scapy/layers/sixlowpan.py:229]: Are these comments needed?

File: scapy/layers/inet6.py [hunk 2]
```diff
@@ -3044,6 +3044,12 @@ class MIP6MH_Generic(_MobilityHeader):  # Mainly for decoding of unknown msg
                    StrLenField("msg", b"\x00" * 2,
                                length_from=lambda pkt: 8 * pkt.len - 6)]
 
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *_, **kargs):
```
Comment (guedou) [inline scapy/layers/inet6.py:3048]: Is this related to this PR?

**Review by guedou — COMMENTED** [review]
Awesome PR! It must have been a long work. I made two minor comments.

**Review by p-l- — APPROVED** [review]
Really cool PR. I'll let @guedou merge it when you have answered his questions.

File: scapy/layers/sixlowpan.py (hunk 1 above)
Comment (gpotter2) [reply scapy/layers/sixlowpan.py:229]: This PR hasn't implemented ICMP and TCP for the HC1 encoding (it's barely mentionned in the RFC, but also very very rarely used and I just couldn't make it work). I just left the comments of where those features should go

File: scapy/layers/inet6.py (hunk 2 above)
Comment (gpotter2) [reply scapy/layers/inet6.py:3048]: I tried at first to properly dissect the HC1 IPv6 extensions but the padding is not bit round (some layers are 20 bits with no padding), so it's not really supported.
This was used to do that. Despite not being used in the end, I thought it still made sense

Comment (guedou) [reply scapy/layers/inet6.py:3048]: If this is not used, I think that it is better to remove this function.

File: scapy/layers/sixlowpan.py (hunk 1 above)
Comment (guedou) [reply scapy/layers/sixlowpan.py:229]: Ok. I understand. Could you add a comment, for example `# TODO: ICMP & TCP not implemented yet for the HC! encoding`, before these commented `PacketField`?
